File: 063_UniquePaths.py
"""
Topic: Dynamic Programming / Combinatorics
Problem: Unique Paths
Description: Number of ways to reach bottom-right in m x n grid.
Time: O(min(m,n)) using combinatorics
Space: O(1)
"""

from math import comb
from typing import List


def uniquePaths(m: int, n: int) -> int:
    # To move (m-1) down and (n-1) right: total moves = m+n-2, choose m-1
    return comb(m + n - 2, m - 1)


if __name__ == '__main__':
    print(uniquePaths(3, 7))  # 28
    print(uniquePaths(3, 2))  # 3

File: 063_UniquePaths.cpp
// Topic: Combinatorics / DP
// Unique Paths: count lattice paths from top-left to bottom-right
// Time Complexity: O(min(m,n)) if using multiplicative formula
// Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;

long long uniquePaths(int m, int n)
{
    int k = min(m - 1, n - 1);
    long double res = 1;
    for (int i = 1; i <= k; i++)
    {
        res = res * (m + n - 2 - k + i) / i;
    }
    return (long long)(res + 0.5L);
}

int main()
{
    cout << uniquePaths(3, 7) << "\n";  // 28
    cout << uniquePaths(3, 2) << "\n";  // 3
    return 0;
}
